cpp/src/stream_compaction/distinct.cu

```diff
@@ -47,8 +35,87 @@
 
 namespace cudf {
 namespace detail {
+
+rmm::device_uvector<size_type> get_distinct_indices(table_view const& input,
+                                                    duplicate_keep_option keep,
+                                                    null_equality nulls_equal,
+                                                    rmm::cuda_stream_view stream,
+                                                    rmm::mr::device_memory_resource* mr)
+{
+  if (input.num_rows() == 0 or input.num_columns() == 0) {
+    return rmm::device_uvector<size_type>(0, stream, mr);
+  }
+
+  auto map = hash_map_type{compute_hash_table_size(input.num_rows()),
+                           cuco::sentinel::empty_key{COMPACTION_EMPTY_KEY_SENTINEL},
+                           cuco::sentinel::empty_value{COMPACTION_EMPTY_VALUE_SENTINEL},
+                           detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+                           stream.value()};
+
+  auto const preprocessed_input =
+    cudf::experimental::row::hash::preprocessed_table::create(input, stream);
+  auto const has_nulls = nullate::DYNAMIC{cudf::has_nested_nulls(input)};
+
+  auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
+  auto const key_hasher = experimental::compaction_hash(row_hasher.device_hasher(has_nulls));
+
+  auto const row_comp  = cudf::experimental::row::equality::self_comparator(preprocessed_input);
+  auto const key_equal = row_comp.equal_to(has_nulls, nulls_equal);
+
+  auto const pair_iter = cudf::detail::make_counting_transform_iterator(
+    size_type{0}, [] __device__(size_type const i) { return cuco::make_pair(i, i); });
+  map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+
+  auto output_indices = rmm::device_uvector<size_type>(map.get_size(), stream, mr);
+
+  // If we don't care about order, just gather indices of distinct keys taken from map.
+  if (keep == duplicate_keep_option::KEEP_ANY) {
+    map.retrieve_all(output_indices.begin(), thrust::make_discard_iterator(), stream.value());
+    return output_indices;
+  }
+
+  // For other keep options, perform a (sparse) reduce-by-row on the rows compared equal.
+  auto const reduction_results = spare_reduce_by_row(map,
+                                                     preprocessed_input,
```

Comment:
```suggestion
                                                     std::move(preprocessed_input),
```
Move to avoid reference counting if passing by value.